src/Sentry/Protocol/Metrics/Metric.cs

```diff
@@ -0,0 +1,119 @@
+
+    public MeasurementUnit? Unit { get; private set; }
+
+    public IDictionary<string, string> Tags { get; private set; }
+
+    public abstract void Add(double value);
+
+    protected abstract void WriteValues(Utf8JsonWriter writer, IDiagnosticLogger? logger);
+
+    public void WriteTo(Utf8JsonWriter writer, IDiagnosticLogger? logger)
+    {
+        writer.WriteStartObject();
+        writer.WriteString("type", GetType().Name);
+        writer.WriteSerializable("event_id", EventId, logger);
+        writer.WriteString("name", Key);
+        writer.WriteString("timestamp", Timestamp);
+        if (Unit.HasValue)
+        {
+            writer.WriteStringIfNotWhiteSpace("unit", Unit.ToString());
+        }
+        writer.WriteStringDictionaryIfNotEmpty("tags", Tags!);
+        WriteValues(writer, logger);
+        writer.WriteEndObject();
+    }
+
+    protected abstract IEnumerable<IConvertible> SerializedStatsdValues();
+
+    public async Task SerializeAsync(Stream stream, IDiagnosticLogger? logger, CancellationToken cancellationToken = default)
+    {
+        /*
+         * We're serializing using the statsd format here: https://github.com/b/statsd_spec
+         */
+        var metricName = MetricHelper.SanitizeKey(Key);
+        await Write($"{metricName}@").ConfigureAwait(false);
+        var unit = Unit ?? MeasurementUnit.None;
+// We don't need ConfigureAwait(false) here as ConfigureAwait on metricName above avoids capturing the ExecutionContext.
+#pragma warning disable CA2007
+        await Write(unit.ToString());
+
+        foreach (var value in SerializedStatsdValues())
+        {
+            await Write($":{value.ToString(CultureInfo.InvariantCulture)}");
+        }
+
+        await Write($"|{StatsdType}");
+
+        if (Tags is { Count: > 0 } tags)
+        {
+            await Write("|#");
+            var first = true;
+            foreach (var (key, value) in tags)
+            {
+                var tagKey = MetricHelper.SanitizeKey(key);
+                if (string.IsNullOrWhiteSpace(tagKey))
+                {
+                    continue;
+                }
+                if (first)
+                {
+                    first = false;
+                }
+                else
+                {
+                    await Write(",");
+                }
+                await Write($"{key}:SanitizeValue(value)");
+            }
+        }
+#pragma warning restore CA2007
+
+        await Write($"|T{Timestamp.GetTimeBucketKey().ToString(CultureInfo.InvariantCulture)}\n").ConfigureAwait(false);
+        return;
+
+        async Task Write(string content)
+        {
+            await stream.WriteAsync(Encoding.UTF8.GetBytes(content), cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    public void Serialize(Stream stream, IDiagnosticLogger? logger)
+    {
+        SerializeAsync(stream, logger).GetAwaiter().GetResult();
+    }
+
+    private string StatsdType =>
```

Comment:
We do? I can't find it.

Reply:
Ah I see... yeah there's one that switches on the `MetricType` enum and another that switches on the Metric class type... potentially we could add an abstract MetricType property to the Metric base class and use that instead of a switch on the class type - then we could get rid of one of these.